django/db/models/fields/related.py

```diff
@@ -731,26 +732,23 @@ def get_reverse_path_info(self):
         pathinfos = [PathInfo(from_opts, opts, (opts.pk,), self.remote_field, not self.unique, False)]
         return pathinfos
 
-    def get_lookup(self, lookup_name):
-        if lookup_name == 'in':
-            return RelatedIn
-        elif lookup_name == 'exact':
-            return RelatedExact
-        elif lookup_name == 'gt':
-            return RelatedGreaterThan
-        elif lookup_name == 'gte':
-            return RelatedGreaterThanOrEqual
-        elif lookup_name == 'lt':
-            return RelatedLessThan
-        elif lookup_name == 'lte':
-            return RelatedLessThanOrEqual
-        elif lookup_name == 'isnull':
-            return RelatedIsNull
-        else:
-            raise TypeError('Related Field got invalid lookup: %s' % lookup_name)
-
-    def get_transform(self, *args, **kwargs):
-        raise NotImplementedError('Relational fields do not support transforms.')
+    def _get_lookup(self, lookup_name):
+        try:
+            return self.class_lookups[lookup_name]
+        except KeyError:
+            # To allow for inheritance, check parent class' class_lookups.
+            for parent in inspect.getmro(self.__class__):
+                # Related lookups need to be explicitly registered to related fields.
+                if parent is ForeignObject:
+                    break
+                if 'class_lookups' not in parent.__dict__:
```

Comment:
So I reran the perf tests on the current master and on a get_lookups() branch w/ caching. 

```
s = "models.DataPoint._meta.get_field('name').get_lookup('in')"
r = "models.RelatedPoint._meta.get_field('data').get_lookup('in')"

Django master:
s result [0.2662360668182373, 0.27753305435180664, 0.2679319381713867, 0.2692220211029053, 0.2688179016113281]
r result [0.058827877044677734, 0.057289838790893555, 0.05742311477661133, 0.05547499656677246, 0.05588197708129883]

get_lookups() w/ cache:
s result [0.3323400020599365, 0.34210205078125, 0.33652591705322266, 0.3279271125793457, 0.3341038227081299]
r result [0.3339991569519043, 0.3370540142059326, 0.33731698989868164, 0.3404569625854492, 0.3389909267425537]

get_lookups():
s result [0.7437207698822021, 0.7416818141937256, 0.7478539943695068, 0.7437272071838379, 0.7346601486206055]
r result [0.589561939239502, 0.5912292003631592, 0.5971081256866455, 0.5899128913879395, 0.5951800346374512]
```

A couple of notes:
- The cached lookups aren't significantly slower, and performance seems consistent across standard fields and related fields. 
- It's worth noting that the related lookups are fast on the current master because they're hardcoded.
- Caching works by adding a `cached_lookups` to the current class `__dict__`. 
- Cache busting works by simply `del`eting the `cached_lookups`.  It's necessary to do this to the subclasses too.
- I don't know if it's acceptable to use `__subclasses__()` in the cache busting method, however I don't know of another way of getting all subclasses for a Field.